core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
     if (sessionTimeoutMs < groupConfig.groupMinSessionTimeoutMs ||
       sessionTimeoutMs > groupConfig.groupMaxSessionTimeoutMs) {
       responseCallback(joinError(memberId, Errors.INVALID_SESSION_TIMEOUT))
-    } else {
+      return
+    }
+
+    if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+      doUnknownJoinGroup(groupId, requireKnownMemberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+      return
+    }
+
+    groupManager.getGroup(groupId) match {
+      case None =>
+        responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
+      case Some(group) =>
+        doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+    }
+  }
+
+  private def protocolNotMatch(group: GroupMetadata,
+                               memberId: String,
+                               protocolType: String,
+                               protocols: List[(String, Array[Byte])],
+                               responseCallback: JoinCallback): Boolean = {
+    val groupProtocolNotSupportedByNewMember = !group.is(Empty) &&
+      (!group.protocolType.contains(protocolType) || !group.supportsProtocols(protocols.map(_._1).toSet))
+    val firstMemberWithEmptyProtocolOrProtocolType = group.is(Empty) && (protocols.isEmpty || protocolType.isEmpty)
+    val notMatch = groupProtocolNotSupportedByNewMember || firstMemberWithEmptyProtocolOrProtocolType
+    if (notMatch) {
+      responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
+    }
+    notMatch
+  }
+
+  private def doUnknownJoinGroup(groupId: String,
+                                 requireKnownMemberId: Boolean,
+                                 clientId: String,
+                                 clientHost: String,
+                                 rebalanceTimeoutMs: Int,
+                                 sessionTimeoutMs: Int,
+                                 protocolType: String,
+                                 protocols: List[(String, Array[Byte])],
+                                 responseCallback: JoinCallback): Unit = {
+    val newMemberId = clientId + "-" + GroupMetadata.generateMemberIdSuffix
+    var group = groupManager.getGroup(groupId).orNull
+    if (group == null) {
       // only try to create the group if the group is not unknown AND
```

Comment:
here we're creating a group if the group is unknown and the member id is unknown, which goes opposite to the comment (saying that the group **should not** be unknown)

Is this a stale comment or am I not interpreting "unknown group" correctly?

Reply:
I think this is stale comment. Let me rephrase it to be correct. You could double check the trunk to see that it actually doesn't make sense (correct way: group is unknown && member id is unknown)